packages/vm/lib/runBlockchain.ts

```diff
@@ -24,50 +23,78 @@ export default function runBlockchain(this: VM, blockchain: Blockchain): Promise
     })
 
     function processBlock(block: any, reorg: boolean, cb: any) {
-      async.series([getStartingState, runBlock], cb)
+      async function _processBlock() {
+        // determine starting state for block run
+        function getStartingState(cb: any) {
+          // if we are just starting or if a chain re-org has happened
+          if (!headBlock || reorg) {
+            blockchain.getBlock(block.header.parentHash, function (err: any, parentBlock: any) {
+              parentState = parentBlock.header.stateRoot
+              // generate genesis state if we are at the genesis block
+              // we don't have the genesis state
+              if (!headBlock) {
+                const generateCanonicalGenesis = callbackify(
+                  self.stateManager.generateCanonicalGenesis.bind(self.stateManager),
+                )
+                return generateCanonicalGenesis(cb)
+              } else {
+                cb(err)
+              }
+            })
+          } else {
+            parentState = headBlock.header.stateRoot
+            cb()
+          }
+        }
 
-      // determine starting state for block run
-      function getStartingState(cb: any) {
-        // if we are just starting or if a chain re-org has happened
-        if (!headBlock || reorg) {
-          blockchain.getBlock(block.header.parentHash, function (err: any, parentBlock: any) {
-            parentState = parentBlock.header.stateRoot
-            // generate genesis state if we are at the genesis block
-            // we don't have the genesis state
-            if (!headBlock) {
-              const generateCanonicalGenesis = callbackify(
-                self.stateManager.generateCanonicalGenesis.bind(self.stateManager),
-              )
-              return generateCanonicalGenesis(cb)
-            } else {
-              cb(err)
-            }
-          })
-        } else {
-          parentState = headBlock.header.stateRoot
-          cb()
+        // run block, update head if valid
+        function runBlock(cb: any) {
+          self
+            .runBlock({
+              block: block,
+              root: parentState,
+            })
+            .then(() => {
+              // set as new head block
+              headBlock = block
+              cb()
+            })
+            .catch((err) => {
+              // remove invalid block
+              blockchain.delBlock(block.header.hash(), function () {
+                cb(err)
+              })
+            })
         }
-      }
 
-      // run block, update head if valid
-      function runBlock(cb: any) {
-        self
-          .runBlock({
-            block: block,
-            root: parentState,
+        await new Promise((resolve, reject) =>
```

Comment:
you can make your life a lot easier using node.js util promisify and callbackify along with async/await syntax to flatten all this code, but no worries i can do that in a follow-up pr (or here if you'd prefer)

after blockchain (+ ethash) are converted the entire monorepo will be promise-based (as i don't think there's any more internal usage of callbacks) so that will be great.

Reply:
I'll clean up all these promises, it makes more sense to do it in this PR 😄